### PR TITLE
Add \file_input_raw:n

### PR DESCRIPTION
**Pull request by josephwright**

### CODE REVIEW

File: l3kernel/l3file.dtx [hunk 1]
```diff
@@ -917,6 +917,21 @@
...
+%   \end{syntax}
+%   Searches for \meta{file name} in the path as detailed for
+%   \cs{file_if_exist:nTF}, and if found reads in the file as
+%   additional \TeX{} source. No data concerning the file is
+%   tracked. If the file is not found, no action is taken.
+%   \begin{texnote}
+%     This function is intended only for contexts where files must
+%     be read purely by expansion, for example at the start of a
+%     table cell in a \tn{halign}.
+%   \end{texnote}
```
Comment (FrankMittelbach) [inline l3kernel/l3file.dtx:932]: In particular, it omits the execution of all file hooks.

or something like that. Also: shouldn't it be  "an \tn{halign}" ?

File: l3kernel/l3file.dtx [hunk 2]
```diff
@@ -3355,6 +3370,24 @@
 % \end{macro}
 % \end{macro}
 %
+% \begin{macro}{\file_input_raw:n, \@@_input:n}
```
Comment (FrankMittelbach) [inline l3kernel/l3file.dtx:3373]: `\@@_input_raw:n` ?

File: l3kernel/l3file.dtx [hunk 3]
```diff
@@ -3355,6 +3370,24 @@
...
+%   No error checking, no tracking.
+%    \begin{macrocode}
+\cs_new:Npn \file_input_raw:n #1
+  { \exp_args:Ne \@@_input_raw:n { \file_full_name:n {#1} } }
+\cs_new:Npx \@@_input_raw:n #1
+  {
+    \exp_not:N \tl_if_blank:nF {#1}
+      {
+        \exp_not:N \tex_input:D
+          \sys_if_engine_luatex:TF
```
Comment (FrankMittelbach) [inline l3kernel/l3file.dtx:3383]: you can run only in one engine for a given run, so why make this a runtime test?

**Review by FrankMittelbach — CHANGES_REQUESTED** [review]
some minor suggestions for improvement

File: l3kernel/l3file.dtx (hunk 3 above)
Comment (josephwright) [reply l3kernel/l3file.dtx:3383]: This is all inside an `x`-type definition precisely for that reason :)

File: l3kernel/l3file.dtx (hunk 1 above)
Comment (muzimuzhi) [reply l3kernel/l3file.dtx:932]: File hooks are LaTeX2e things so I think for expl3 functions there're always no file hooks involved?

Comment (josephwright) [reply l3kernel/l3file.dtx:932]: I thin the h' in `halign` is voiced, so 'a' not 'an' (so like 'hotel' not 'hour')

Comment (car222222) [reply l3kernel/l3file.dtx:932]: Not around my manor it ain't!  Look it up??:-)

It is in fact neither voiced nor unvoiced as, in my part of TeXland, it is not even pronounced!--
halign is voiced as aitch-align
hence "an halign", please!

Comment (car222222) [reply l3kernel/l3file.dtx:932]: Not around my manor it ain't!  Look it up??:-)

It is in fact neither voiced nor unvoiced as, in my part of TeXland, it is not even pronounced!--
halign is voiced as aitch-align
hence "an halign", please!

Comment (car222222) [reply l3kernel/l3file.dtx:932]: Joseph, given his profession, may be confusing it "haline"??

**Comment by josephwright** [comment]
I take it everyone is happy with the suggested name?

File: l3kernel/l3file.dtx [hunk 4]
```diff
@@ -3355,6 +3370,24 @@
...
+            { {#1} }
+            { \exp_not:N \__kernel_file_name_quote:n {#1} \scan_stop: }
+        }
```
Comment (davidcarlisle) [inline l3kernel/l3file.dtx:3386]: Can we not (yet) assume a   primitive `\input` takes  braces? (so lose the luatex test?) (needs texlive 2020 or later)

File: l3kernel/l3file.dtx [hunk 5]
```diff
@@ -917,6 +917,21 @@
...
 %
+% \begin{function}[added = 2023-05-18, EXP]{\file_input_raw:n}
+%   \begin{syntax}
+%     \cs{file_input_raw:n} \Arg{file name}
+%   \end{syntax}
+%   Searches for \meta{file name} in the path as detailed for
+%   \cs{file_if_exist:nTF}, and if found reads in the file as
+%   additional \TeX{} source. No data concerning the file is
+%   tracked. If the file is not found, no action is taken.
+%   \begin{texnote}
```
Comment (davidcarlisle) [inline l3kernel/l3file.dtx:928]: > no action is taken.

sounds like it is skipped (like `\@input`) but I think you mean "no action is taken _by latex_" so  you get an engine level file not found error.

File: l3kernel/l3file.dtx (hunk 4 above)
Comment (josephwright) [reply l3kernel/l3file.dtx:3386]: I wondered about that: I think we can't yet at like you say it's TL'20, and even moving to TL'19 as a requirement is quite 'up to date'!

File: l3kernel/l3file.dtx (hunk 5 above)
Comment (josephwright) [reply l3kernel/l3file.dtx:928]: Er, not: we have at this point applied `\file_full_name:n` and got an empty result, so we branch to using `\tl_if_blank:nF` and so a missing file will be entirely skipped.

Comment (davidcarlisle) [reply l3kernel/l3file.dtx:928]: oops, sorry I should have noticed `full_name_` did that.

This seems wrong, normalising the name  is  needed for assigning the correct hook if you have hooks, but here to be honest  I think I'd just let it to the primitive, or at most define it as `\def\file_input_raw:n#1{\@@input#1\relax}` so braces work on old releases.

The main use case here is to underlie a document level input command that can be used in tables, but if this one skips missing file errors, I think any document level command would have to not use this, and be defined directly from the primitive.

Comment (josephwright) [reply l3kernel/l3file.dtx:928]: I've adjusted to issue an expandable error - so something we still control

**Pull request merged**